### PR TITLE
Add an option for the Visvalingam simplification algorithm

### PR DESCRIPTION
**Pull request by e-n-f**
Adds a new `--visvalingam` option, to use the Visvalingam simplification algorithm rather than the Douglas-Peucker algorithm.

### CODE REVIEW

**Review by migurski — CHANGES_REQUESTED** [review]
Looks good! I put two suggestions in comments, otherwise good to go and I am thrilled that we can benefit from @paulmach’s work

File: geometry.cpp [hunk 1]
```diff
@@ -864,6 +875,7 @@ drawvec simplify_lines(drawvec &geom, int z, int detail, bool mark_tile_bounds,
 	}
 
 	return out;
+	return out;
```
[the remark this answers is not in the document]
Comment (e-n-f) [reply geometry.cpp:878]: oops, yes!

File: version.hpp [hunk 2]
```diff
@@ -1,6 +1,6 @@
 #ifndef VERSION_HPP
 #define VERSION_HPP
 
-#define VERSION "v2.6.4"
+#define VERSION "v2.6.5"
```
[the remark this answers is not in the document]
Comment (e-n-f) [reply version.hpp:4]: Good point!

File: geometry.cpp [hunk 3]
```diff
@@ -849,8 +849,19 @@ drawvec simplify_lines(drawvec &geom, int z, int detail, bool mark_tile_bounds,
 			geom[i].necessary = 1;
 			geom[j - 1].necessary = 1;
 
+			// empirical mapping from douglas-peucker simplifications
+			// to visvalingam simplifications that yield similar
+			// output sizes
+			double sim = simplification * (0.1596 * z + 0.878);
```
[the remark this answers is not in the document]
Comment (e-n-f) [reply geometry.cpp:855]: The second adjustment is from this set of experiments with existing `--simplification` values and the resulting output size for `ne_10m_admin_0_countries` at z0, compared to sqrt(visvalingam area * zoom adjustment) that yields approximately the same output size:

existing -S | output size | visvalingam area | output size | sqrt(area * zoomscale) | output size
----------- | ----------- | --------------- | ----------- | --------------------
--- | ------
0.5 | 3527690 | 1000000 | 3085053 | 0.9 | 3085049
1 | 2168684 | 2000000 | 2225159 | 1.4 | 2225155
2 | 1544182 | 6000000 | 1557178 | 2.4 | 1557174
4 | 1231908 | 16000000 | 1285993 | 4.6 | 1230857
8 | 1084641 | 80000000 | 1086389 | 9.3 | 1078328
16 | 1016324 | 300000000 | 1016803 | 17.5 | 1015243
32 | 973557 | 2000000000 | 973644 | 44 | 973589
64 | 961327 | 15900000000 | 960927 | 123 | 960919

```
gnuplot> stats "log" using (log($1)):(log($9))
  Linear Model:       y = 1.002 x + 0.3043
```

The first adjustment then comes from doing the same at multiple zoom levels to see what Visvalingam `-S` with the previous adjustment yields about the same output size as a constant Douglas-Peucker `-S`:

zoom level | existing -S | visvalingam -S
---------- | ---------- | ----------
5 | 4 | 7
4 | 4 | 6
3 | 4 | 5.2
2 | 4 | 4.55
1 | 4 | 4.1
0 | 4 | 3.8

```
gnuplot> stats "vis-scaling" using 1:($5 / $3)
  Linear Model:       y = 0.1596 x + 0.878
```

I am not exactly thrilled with this process but can't think of a cleaner way to relate the threshold parameters for the two algorithms.

[the remark this answers is not in the document]
Comment (e-n-f) [reply geometry.cpp:855]: Looking at this again now, I clearly messed up on reproducing the `-S 0.5` size.

**Comment by bdon** [comment]
Nice! Is this more time-consuming than Douglas-Peucker?

**Comment by e-n-f** [comment]
@bdon Yes, Douglas-Peucker is faster.

**Pull request merged**